### PR TITLE
PR for Lesson09 Made some improvements to mailroom_oo

### PR DESCRIPTION
**Pull request by LouReis**
Hello Natasha,

I made some improvements but have a few issues with mailroom_oo.py:

-Attempt to use "find_donor" method but is says Donor doesn't have 'name' attribute (but it does)
-My get_report is not sorted, is there an easy way to do this with class objects?  I cannot use the sorted function as I did with dictionaries.

Thanks,
Lou

### CODE REVIEW

**Comment by natasha-aleksandrova** [comment]
there are various examples online on how to sort objects, here is one:
https://www.pythoncentral.io/how-to-sort-a-list-tuple-or-object-with-sorted-in-python/

**Comment by natasha-aleksandrova** [comment]
this is your `find_donor` function:
```
    def find_donor(self, name):
        """
        find a donor in the donor db
        :param: the name of the donor
        :returns: The donor data structure -- None if not in the self.donor_data
        """
        return self.donor_data.get(Donor.name(name))
```

`Donor.name` tries to access class variable `name` which does not exist
`Donor().name` tries to access instance attribute `name` which does exist
`self.donor_data` holds reference to all your donor objects - you should _only_ use this for all of your data access:
```
def find_donor(self, name):
    self.donor_data.get(name)  # this will return Donor object or None
```

File: students/LouReis/Lesson09/mailroom_oo.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    def list_donors(self):
+        """
+        creates a list of the donors as a string, so they can be printed
+        Not calling print from here makes it more flexible and easier to
+        test
+        """
+        listing = ["Donor list:"]
+        for donor in self.donors:
+            listing.append(donor.name)
+        return "\n".join(listing)
```
Comment (natasha-aleksandrova) [inline students/LouReis/Lesson09/mailroom_oo.py]: `self.donors` doesn't exist.
cleaner versions:
`return ["Donor list:"] + list(self.donor_data)`

File: students/LouReis/Lesson09/mailroom_oo.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+
+    def add_donor(self, name):
+        """
+        Add a new donor to the donor db
+        :param: the name of the donor
+        :returns: the new Donor data structure
+        """
+        donor = Donor(name)
+        self.donor_data[donor.name] = donor
+        return donor
```
Comment (natasha-aleksandrova) [inline students/LouReis/Lesson09/mailroom_oo.py]: you shouldn't have a return in this function

File: students/LouReis/Lesson09/mailroom_oo.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+        :returns: string with letter
+        note: This doesn't actually write to a file -- that's a separate
+              function. This makes it more flexible and easier to test.
+        """
+        return dedent('''Dear {0:s},
+              Thank you for your very kind donation of ${1:.2f}.
+              It will be put to very good use.
+                             Sincerely,
+                                -The Team
+              '''.format(donor.name)
```
Comment (natasha-aleksandrova) [inline students/LouReis/Lesson09/mailroom_oo.py]: this needs to go into `Donor` class.

File: students/LouReis/Lesson09/mailroom_oo.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+    donation = 0
+    if donor in Donor:
+        print("You have entered an existing donor:", donor)
+        try:
+            donation = float(input("Please enter the donation amount '0.00':"))
+        except ValueError:
+            print("\n\n----------You have entered an invalid value, returning to Main Menu----------\n\n")
+            main_menu(main_prompt,menu_options_dict)
+        temp = Donor(donor)
+        temp.add_donation(donation)
```
Comment (natasha-aleksandrova) [inline students/LouReis/Lesson09/mailroom_oo.py]: this needs to be in DonorDb class

**Comment by LouReis** [comment]
Thanks for the feedback, I appreciate it...I'll review & update accordingly...

 

Lou

 

Sent: Saturday, January 19, 2019 at 6:46 PM

**Pull request merged**